drivers/sensor/icm42688/icm42688_emul.c

```diff
@@ -117,9 +118,298 @@ static const struct spi_emul_api icm42688_emul_spi_api = {
+}
+
+/**
+ * @brief Get current full-scale gyro range in milli-degrees per second based on register config,
+ *        along with corresponding sensitivity and shift. See datasheet section 3.1, table 1.
+ */
+static void icm42688_emul_get_gyro_settings(const struct emul *target, int *fs_mdps,
+					    int *sensitivity, int8_t *shift)
+{
+	uint8_t reg;
+
+	int sensitivity_out, fs_mdps_out;
+	int8_t shift_out;
+
+	icm42688_emul_get_reg(target, REG_GYRO_CONFIG0, &reg, 1);
+
+	switch ((reg & MASK_GYRO_UI_FS_SEL) >> 5) {
+	case BIT_GYRO_UI_FS_2000:
+		/* Milli-degrees per second */
+		fs_mdps_out = 2000000;
+		/* 10x LSBs/deg/s */
+		sensitivity_out = 164;
+		/* Shifts are based on rad/s: `(fs_mdps * pi / 180 / 1000)` */
+		shift_out = 6; /* +/- 34.90659 */
+		break;
+	case BIT_GYRO_UI_FS_1000:
+		fs_mdps_out = 1000000;
+		sensitivity_out = 328;
+		shift_out = 5; /* +/- 17.44444 */
+		break;
+	case BIT_GYRO_UI_FS_500:
+		fs_mdps_out = 500000;
+		sensitivity_out = 655;
+		shift_out = 4; /* +/- 8.72222 */
+		break;
+	case BIT_GYRO_UI_FS_250:
+		fs_mdps_out = 250000;
+		sensitivity_out = 1310;
+		shift_out = 3; /* +/- 4.36111 */
+		break;
+	case BIT_GYRO_UI_FS_125:
+		fs_mdps_out = 125000;
+		sensitivity_out = 2620;
+		shift_out = 2; /* +/- 2.18055 */
+		break;
+	case BIT_GYRO_UI_FS_62_5:
+		fs_mdps_out = 62500;
+		sensitivity_out = 5243;
+		shift_out = 1; /* +/- 1.09027 */
+		break;
+	case BIT_GYRO_UI_FS_31_25:
+		fs_mdps_out = 31250;
+		sensitivity_out = 10486;
+		shift_out = 0; /* +/- 0.54513 */
+		break;
+	case BIT_GYRO_UI_FS_15_625:
+		fs_mdps_out = 15625;
+		sensitivity_out = 20972;
+		shift_out = -1; /* +/- 0.27256 */
+		break;
+	default:
+		__ASSERT_UNREACHABLE;
+	}
+
+	if (fs_mdps) {
+		*fs_mdps = fs_mdps_out;
+	}
+	if (sensitivity) {
+		*sensitivity = sensitivity_out;
+	}
+	if (shift) {
+		*shift = shift_out;
+	}
+}
+
+/**
+ * @brief Helper function for calculating gyroscope ranges. Considers the current full-scale
+ *        register config
+ */
+static void icm42688_emul_get_gyro_ranges(const struct emul *target, q31_t *lower, q31_t *upper,
+					  q31_t *epsilon, int8_t *shift)
+{
+	/* millidegrees/second */
+	int fs_mdps;
+	/* 10x LSBs per degrees/second*/
+	int sensitivity;
+
+	icm42688_emul_get_gyro_settings(target, &fs_mdps, &sensitivity, shift);
+
+	/* Reduce the actual range of gyroscope values. Some full-scale ranges actually exceed the
+	 * size of an int16 by a small margin. For example, FS_SEL=0 has a +/-2000 deg/s range with
+	 * 16.4 bits/deg/s sensitivity (Section 3.1, Table 1). This works out to register values of
+	 * +/-2000 * 16.4 = +/-32800. This will cause the expected value to get clipped when
+	 * setting the register and throw off the actual reading. Therefore, scale down the range
+	 * to 99% to avoid the top and bottom edges.
+	 */
+
+	fs_mdps *= 0.99;
+
+	/* Epsilon is equal to 1.5 bit-counts worth of error. */
+	*epsilon = (3 * SENSOR_PI * Q31_SCALE * 10LL / 1000000LL / 180LL / sensitivity / 2LL) >>
+		   *shift;
```

Comment:
@tristan-google  In case shift==-1 (BIT_GYRO_UI_FS_15_625 case in icm42688_emul_get_gyro_settings), the right shift operation has an undefined behavior

> The result of E1 >> E2 is E1 right-shifted E2 bit positions. If E1 has an unsigned type
or if E1 has a signed type and a nonnegative value, the value of the result is the integral
part of the quotient of E1 divided by the quantity, 2 raised to the power E2. **If E1 has a
signed type and a negative value, the resulting value is implementation-defined**